lib/src/license_detection/new_tokenizer.dart

```diff
@@ -0,0 +1,278 @@
+  /// Any Leading or standalone punctuation form a separate token
+  /// Example `! !hello!` --> `!`, `!`, `hello!`
+  while (!_scanner.isDone) {
+    var prevState = _scanner.state;
+    var char = _scanner.readChar();
+
+    // Check if the character is alphanumeric,
+    // If it is create a span until you reach space or new Line.
+    // If not create a span for of the single character.
+    // This approach will help us to deal with comment indicators, bullets
+    // and list items.
+    if (_isAlphaNumeric(char)) {
+      while (!_scanner.isDone) {
+        char = _scanner.readChar();
+        if (_isSpace(char) || char == 10) {
+          break;
+        }
+      }
+
+      // Move the scanner back to exclude space or newLine character
+      // from token.
+      _scanner.position--;
+      _addNewToken(_scanner.spanFrom(prevState), tokenID++, tokens);
+
+      prevState = _scanner.state;
+      _scanner.position++;
+
+      if (_isNewLine(char)) {
+        _addNewToken(_scanner.spanFrom(prevState), tokenID++, tokens);
+      }
+    } else if (!_isSpace(char)) {
+      _addNewToken(_scanner.spanFrom(prevState), tokenID++, tokens);
+    }
+  }
+
+  /// Normalizes tokens using the approach of [google licenseClassifier]
+  /// to provide better chances of matching.
+  /// [google licenseClassifier]: https://github.com/google/licenseclassifier/blob/bb04aff29e72e636ba260ec61150c6e15f111d7e/v2/tokenizer.go#L34
+  tokens = _cleanNewTokens(tokens);
+
+  return tokens;
+}
+
+void _addNewToken(FileSpan fileSpan, int tokenID, List<NewToken> tokens) {
+  final normText = fileSpan.text.toLowerCase();
+  final tok = NewToken(normText, tokenID++, fileSpan.start.line, fileSpan);
+  tokens.add(tok);
+}
+
+List<NewToken> _cleanNewTokens(List<NewToken> tokens) {
+  var output = <NewToken>[];
+  var tokenID = 0;
+  var firstInLine = true;
+
+  for (var token in tokens) {
+    // Ignore new line tokens for now.
+    // If accuracy of detection is low apply
+    // Guideline 2.1.4: Text that can be omited from license.
+    if (token.normalizedText == '\n') {
+      firstInLine = true;
+      continue;
+    }
+
+    var char = token.normalizedText.codeUnits.first;
+
+    // Ignores single puntcuations as they are
+    // not significant in detection.
+    if (!_isAlphaNumeric(char) && token.normalizedText.length == 1) {
+      continue;
+    }
+
+    // Ignores list items.
+    if (firstInLine && _isListItem(token.normalizedText)) {
+      continue;
+    }
+
+    firstInLine = false;
+
+    final text = _cleanToken(token.normalizedText);
+
+    output.add(NewToken(text, tokenID++, token.line, token.fileSpan));
+  }
+
+  return output;
+}
+
+/// Normalizes the tokens using the approach of [google licenseClassifier]
+/// to provide better chances of matching.
+/// [google licenseClassifier]  https://github.com/google/licenseclassifier/blob/bb04aff29e72e636ba260ec61150c6e15f111d7e/v2/tokenizer.go#L85
+String _cleanToken(String tok) {
```

Comment:
This looks very far from: https://github.com/google/licenseclassifier/blob/bb04aff29e72e636ba260ec61150c6e15f111d7e/v2/tokenizer.go#L85

Try to decipher what kind of tokens you want, then we can figure out how to create them... For example, if we want to allow various forms of hyphenation (dashes like `-`) inside of words.. then maybe we should just expand the `_alphaNumericPattern` or add a new pattern of handling such cases before we match with `_alphaNumericPattern`.

Reply:
I can see that it makes sense to make newline tokens, and then handle removal of various forms of bullet points and enumerations after tokenization. But for most of the other kinds of normalization, it might be just as fine to do it as part of the tokenization itself.. but making specific patterns to handle specific cases.